api/users/models.py

```diff
@@ -469,6 +469,33 @@ class HubspotTracker(models.Model):
         related_name="hubspot_tracker",
         on_delete=models.CASCADE,
     )
-    hubspot_cookie = models.CharField(unique=True, max_length=100, null=False)
+    hubspot_cookie = models.CharField(
+        unique=True,
+        max_length=100,
+        null=True,
+        blank=True,
+    )
+    utm_source = models.CharField(max_length=256, null=True, blank=True)
+    utm_medium = models.CharField(max_length=256, null=True, blank=True)
+    utm_campaign = models.CharField(max_length=256, null=True, blank=True)
+    utm_term = models.CharField(max_length=256, null=True, blank=True)
+    utm_content = models.CharField(max_length=256, null=True, blank=True)
+
     created_at = models.DateTimeField(auto_now_add=True)
     updated_at = models.DateTimeField(auto_now=True)
+
+    @staticmethod
+    def build_utm_data(utm_dict: dict[str, str] | None) -> dict[str, str]:
+        if not utm_dict:
+            return {}
+        return {
+            key: utm_value
+            for key, utm_value in utm_dict.items()
+            if key in ALLOWED_UTM_KEYS and utm_value
+        }
+
+    @property
+    def utm_data(self) -> dict[str, str]:
+        return {
+            key: getattr(self, key) for key in ALLOWED_UTM_KEYS if getattr(self, key)
```

Comment:
Uh oh, we're also doing some of that same logic here!

Reply:
Ok, I removed the duplication. I'll answer here concerning the serializer. I didn't go with one as the utms are sent as individual values, and in the DB they are stored in the same way (to avoid JSON manipulation).
So it looked like doing so in the serializer would (from what I see) add an extra formatting step?

`ALLOWED_UTM_KEYS` has 2 purposes, filtering and providing the keys to get the correct attributes.

The way I see it, without duplicated logic, is that:
- the @property acts as the getter from the model
- the @static is more of a helper to build a dict from unstructured data (maybe this one could be moved)

Let me know what you think in the new version or if it's worth using a serializer